src/box/txn.h

```diff
@@ -581,6 +581,26 @@ txn_check_can_continue(struct txn *txn)
 	return 0;
 }
 
+/**
+ * Checks if TCL statements can be executed in the given transaction.
+ * Returns 0 if true. Otherwise, sets diag and returns -1.
+ */
+static inline int
+txn_check_can_tcl(struct txn *txn)
+{
+	enum txn_status status = txn->status;
+	if (status == TXN_ABORTED && txn_has_flag(txn, TXN_IS_ROLLED_BACK)) {
+		/* Cannot execute TCL in already rolled back transaction. */
+		diag_set(ClientError, ER_TXN_ROLLBACK);
+		return -1;
+	} else if (status == TXN_COMMITTED) {
+		/* Cannot execute TCL in already committed transaction. */
+		diag_set(ClientError, ER_TXN_COMMIT);
+		return -1;
+	}
+	return 0;
```

Comment:
I guess the main point is that we should not allow double `commit` or `rollback`.

Reply:
> Now TCL statements related to savepoints use txn_check_continue because creating and rolling back to savepoint is about continuing the transaction rather than its completion.

Well, my plan has failed. The problem is SQL can use rollback to savepoint on a conflicted transaction. Actually, TCL statements related to savepoints don't create any statements, so `txn_check_can_complete` is sufficient there.